### PR TITLE
Azure: add option to force use of CLI credential

### PR DESCRIPTION
**Pull request by letmaik**
What does this PR change? What problem does it solve?
-----------------------------------------------------

In Azure, VMs can have multiple identities at the same time, for example a managed identity and an Azure CLI identity. Sometimes the managed identity is not under control of the user but the user can still login with the Azure CLI. In those cases, being able to use the Azure CLI identity with restic makes sense.

`DefaultAzureCredential` first tries environment variables, managed identity, workload identity, and eventually Azure CLI identity. This PR introduces a new environment variable that forces use of the Azure CLI identity:

```
export AZURE_FORCE_CLI_CREDENTIAL=true
```

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
  - [x] Manually tested. 
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/backend/azure/azure.go [hunk 1]
```diff
@@ -101,6 +101,18 @@ func open(cfg Config, rt http.RoundTripper) (*Backend, error) {
 		if err != nil {
 			return nil, errors.Wrap(err, "NewAccountSASClientFromEndpointToken")
 		}
+	} else if cfg.ForceCliCredential {
+		debug.Log(" - using AzureCLICredential")
+
+		cred, err := azidentity.NewAzureCLICredential(nil)
```
Comment (MichaelEischer) [inline internal/backend/azure/azure.go:107]: The only difference to the `- using DefaultAzureCredential` case is how the credentials are obtained. Please merge both cases.

File: doc/030_preparing_a_new_repo.rst [hunk 2]
```diff
@@ -550,17 +550,23 @@ For authentication export one of the following variables:
...
 
-For authentication using ``az login`` set the resource group name and ensure the user has
-the minimum permissions of the role assignment ``Storage Blob Data Contributor`` on Azure RBAC.
+For authentication using ``az login`` ensure the user has
+the minimum permissions of the role assignment ``Storage Blob Data Contributor`` on Azure RBAC
+for the storage account.
 
 .. code-block:: console
 
-    $ export AZURE_RESOURCE_GROUP=<RESOURCE_GROUP_NAME>
```
Comment (MichaelEischer) [inline doc/030_preparing_a_new_repo.rst:558]: What's the reason for this change?

File: internal/backend/azure/config.go [hunk 3]
```diff
@@ -73,6 +74,8 @@ func (cfg *Config) ApplyEnvironment(prefix string) {
 		cfg.AccountSAS = options.NewSecretString(os.Getenv(prefix + "AZURE_ACCOUNT_SAS"))
 	}
 
+	cfg.ForceCliCredential = os.Getenv(prefix+"AZURE_FORCE_CLI_CREDENTIAL") == "true"
```
Comment (MichaelEischer) [inline internal/backend/azure/config.go:77]: Please use `strconv.ParseBool` for parsing the environment variable. I'd prefer to set `ForceCliCredential` only if the environment variable evaluates to true. (Basically the boolean analogue to the string parameters are currently handled).

File: doc/030_preparing_a_new_repo.rst [hunk 4]
```diff
@@ -550,17 +550,23 @@ For authentication export one of the following variables:
...
 
-Alternatively, if run on Azure, restic will automatically uses service accounts configured
+Alternatively, if run on Azure, restic will automatically use service accounts configured
 via the standard environment variables or Workload / Managed Identities.
 
+To enforce the use of the Azure CLI crendential when other credentials are present, set the following environment variable:
+
+.. code-block:: console
+
+    $ export AZURE_FORCE_CLI_CREDENTIAL=true
```
Comment (MichaelEischer) [inline doc/030_preparing_a_new_repo.rst:568]: There's an environment variable list in `040_backup.rst` please also extend that.

File: doc/030_preparing_a_new_repo.rst [hunk 5]
```diff
@@ -550,17 +550,23 @@ For authentication export one of the following variables:
...
 .. code-block:: console
 
-    $ export AZURE_RESOURCE_GROUP=<RESOURCE_GROUP_NAME>
     $ az login
 
-Alternatively, if run on Azure, restic will automatically uses service accounts configured
+Alternatively, if run on Azure, restic will automatically use service accounts configured
 via the standard environment variables or Workload / Managed Identities.
 
+To enforce the use of the Azure CLI crendential when other credentials are present, set the following environment variable:
```
Comment (MichaelEischer) [inline doc/030_preparing_a_new_repo.rst:564]: Typo `crendential`

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Adding the environment variable seems ok, I didn't find any easy workarounds to ignore the managed identity.

I have a few comments though, see below.

File: internal/backend/azure/azure.go (hunk 1 above)
Comment (letmaik) [reply internal/backend/azure/azure.go:107]: I don't have much experience in Go yet, I was trying this but it fails with "cred declared and not used". Any idea?

```go
	} else {
		var cred *azcore.TokenCredential
		if cfg.ForceCliCredential {
			debug.Log(" - using AzureCLICredential")
			cred, err := azidentity.NewAzureCLICredential(nil)
			if err != nil {
				return nil, errors.Wrap(err, "NewAzureCLICredential")
			}
		} else {
			debug.Log(" - using DefaultAzureCredential")
			cred, err := azidentity.NewDefaultAzureCredential(nil)
			if err != nil {
				return nil, errors.Wrap(err, "NewDefaultAzureCredential")
			}
		}

		client, err = azContainer.NewClient(url, *cred, opts)
		if err != nil {
			return nil, errors.Wrap(err, "NewClient")
		}
	}
```

File: doc/030_preparing_a_new_repo.rst (hunk 2 above)
Comment (letmaik) [reply doc/030_preparing_a_new_repo.rst:558]: This environment variable is unused and has no effect.

File: doc/030_preparing_a_new_repo.rst (hunk 4 above)
Comment (letmaik) [reply doc/030_preparing_a_new_repo.rst:568]: Done.

File: internal/backend/azure/config.go (hunk 3 above)
Comment (letmaik) [reply internal/backend/azure/config.go]: Done.

File: internal/backend/azure/azure.go (hunk 1 above)
Comment (MichaelEischer) [reply internal/backend/azure/azure.go]: I've made the necessary change.

 `:=` declares a new variable within the current scope. That is `cred` in `var cred ...` and `cred, err := ...` are different variables.

**Review by MichaelEischer — APPROVED** [review]
LGTM. I've rebased the PR and added a commit to deduplicate the CLI and default credentials case.

**Comment by letmaik** [comment]
@MichaelEischer Any chance to cut a new release with this? Thanks! :)

**Comment by letmaik** [comment]
@MichaelEischer Sorry to be a pain, do you think 0.17.0 will be out by end of next week? We're rolling out an internal policy that will prevent key-based access to storage accounts and without this PR we won't be able to use restic anymore for certain Azure VM scenarios.

**Comment by MichaelEischer** [comment]
> Sorry to be a pain, do you think 0.17.0 will be out by end of next week?

Definitely not before end of next week, a release at the end of next week might be possible, but is somewhat unlikely. There are a few more changes necessary for the `restore` command before we can cut a new release and I didn't make as much progress as I hoped for.

**Comment by rawtaz** [comment]
@letmaik Can you please use the latest master build instead, temporarily until the new version is out? It's at https://beta.restic.net/latest/ .

**Comment by MichaelEischer** [comment]
I've noticed that cherry-picking the PR for a 0.16.5 patch release works. So we can include the change in a release earlier than expected.

**Comment by MichaelEischer** [comment]
@letmaik 0.16.5 has been released: https://github.com/restic/restic/releases/tag/v0.16.5

**Comment by letmaik** [comment]
@MichaelEischer Thanks a lot, that's extremely helpful! :)
